### PR TITLE
Issues/310 document reference

### LINKED ISSUE #310: Implement DocumentReference Resource
We should implement the DocumentReference resource to represent meta information of Binary resources. A use case would be to provide a public-key for encryption of data to be transmitted between organizations.

[http://www.hl7.org/fhir/documentreference.html](http://www.hl7.org/fhir/documentreference.html)

Important would be to have the search parameter for `author` that allows searching DocumentReferences based on an Organizations identifier.

### PR DESCRIPTION
**Pull request by hhund**
* Adds support for DocumentReference resources
* Adds DocumentReference.​masterIdentifier / DocumentReference.​identifier search parameter
* Adds support for reference resolution of Attachment properties used in Library and DocumentReference
* Some code refactoring

closes #310

### CODE REVIEW

File: dsf-fhir/dsf-fhir-server/src/main/java/org/highmed/dsf/fhir/dao/provider/DaoProviderImpl.java [hunk 1]
```diff
@@ -359,7 +372,7 @@ public ValueSetDao getValueSetDao()
 	public <R extends Resource> Optional<? extends ResourceDao<R>> getDao(Class<R> resourceClass)
 	{
 		@SuppressWarnings("unchecked")
-		ResourceDao<R> value = (ResourceDao<R>) daosByResourecClass.get(resourceClass);
+		ResourceDao<R> value = (ResourceDao<R>) daosByResoureClass.get(resourceClass);
```
Comment (wetret) [inline dsf-fhir/dsf-fhir-server/src/main/java/org/highmed/dsf/fhir/dao/provider/DaoProviderImpl.java:375]: Typo: change to `daosByResourceClass`?

**Review by wetret — APPROVED** [review]
There are some minor typos that could still be fixed, see comments.

File: dsf-fhir/dsf-fhir-rest-adapter/src/main/java/org/highmed/dsf/fhir/service/ReferenceExtractorImpl.java [hunk 2]
```diff
@@ -77,6 +83,11 @@
 		return rel -> new ResourceReference(relatedArtifactLocation, rel);
 	}
 
+	private Function<Attachment, ResourceReference> toResourceReferenceFromAttachment(String relatedArtifactLocation)
```
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-fhir/dsf-fhir-rest-adapter/src/main/java/org/highmed/dsf/fhir/service/ReferenceExtractorImpl.java]: fixed via [ecad59f](https://github.com/highmed/highmed-dsf/pull/333/commits/ecad59fcfeecc95b1b25bf89c5ccc7a5066fbed0)

File: dsf-fhir/dsf-fhir-rest-adapter/src/main/java/org/highmed/dsf/fhir/service/ReferenceExtractorImpl.java [hunk 3]
```diff
@@ -385,6 +442,48 @@ else if (resource instanceof DomainResource)
...
+				"DocumentReference.subject", Patient.class, Practitioner.class, Group.class, Device.class);
+		var author = getReferences(resource, DocumentReference::hasAuthor, DocumentReference::getAuthor,
+				"DocumentReference.author", Practitioner.class, PractitionerRole.class, Organization.class,
+				Device.class, Patient.class, RelatedPerson.class);
+		var authenticator = getReference(resource, DocumentReference::hasAuthenticator,
+				DocumentReference::getAuthenticator, "DocumentReference.authenticator", Practitioner.class,
+				PractitionerRole.class, Organization.class);
+		var custodian = getReference(resource, DocumentReference::hasCustodian, DocumentReference::getCustodian,
+				"DocumentReference.custodian", Organization.class);
+		var relatesTosTarget = getBackboneElementsReference(resource, DocumentReference::hasRelatesTo,
```
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-fhir/dsf-fhir-rest-adapter/src/main/java/org/highmed/dsf/fhir/service/ReferenceExtractorImpl.java]: fixed via [ecad59f](https://github.com/highmed/highmed-dsf/pull/333/commits/ecad59fcfeecc95b1b25bf89c5ccc7a5066fbed0)

File: dsf-fhir/dsf-fhir-rest-adapter/src/main/java/org/highmed/dsf/fhir/service/ReferenceExtractorImpl.java [hunk 4]
```diff
@@ -385,6 +442,48 @@ else if (resource instanceof DomainResource)
...
+				DocumentReference::getRelatesTo, DocumentReferenceRelatesToComponent::hasTarget,
+				DocumentReferenceRelatesToComponent::getTarget, "DocumentReference.relatesTo.target",
+				DocumentReference.class);
+		var contextEncounters = getBackboneElementReferences(resource, DocumentReference::hasContent,
+				DocumentReference::getContext, DocumentReferenceContextComponent::hasEncounter,
+				DocumentReferenceContextComponent::getEncounter, "DocumentReference.context.encounter", Encounter.class,
+				EpisodeOfCare.class);
+		var contextsourcePatientInfo = getBackboneElementReference(resource, DocumentReference::hasContent,
```
[the remark this answers is not in the document]
Comment (hhund) [reply dsf-fhir/dsf-fhir-rest-adapter/src/main/java/org/highmed/dsf/fhir/service/ReferenceExtractorImpl.java]: fixed via [ecad59f](https://github.com/highmed/highmed-dsf/pull/333/commits/ecad59fcfeecc95b1b25bf89c5ccc7a5066fbed0)

File: dsf-fhir/dsf-fhir-server/src/main/java/org/highmed/dsf/fhir/dao/provider/DaoProviderImpl.java (hunk 1 above)
Comment (hhund) [reply dsf-fhir/dsf-fhir-server/src/main/java/org/highmed/dsf/fhir/dao/provider/DaoProviderImpl.java]: fixed via [ecad59f](https://github.com/highmed/highmed-dsf/pull/333/commits/ecad59fcfeecc95b1b25bf89c5ccc7a5066fbed0)

**Pull request merged**